Make specific objects serialize into own files

rainman110 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

This PR introduces a complete and symmetric solution for handling **linked XML objects** in GTlab project files.  
It implements the full round-trip:

- Writing project documents with `<object>` and `aslink="true|1|refonly"`
- Extracting linked objects into external `*.gtobj.xml` files
- Replacing them with `<objectref>` placeholders
- Loading the master document again and **recursively** resolving all `<objectref>`s back into fully inlined `<object>` nodes

Closes #1419

## Fulfilled Acceptance Criteria

 - [x] Objects marked for external serialization are saved into their own XML files.
 - [x] Parent module file contains instead of a full object subtree.
 - [x] Loading a module automatically resolves all external references.
 - [x] All saving operations are atomic across multiple files.
 - [x] Feature works recursively in hierarchical object structures.
 - [x] Unresolved linked files result in dummy objects
 - [x] Unresolved files are not replaced when saving
 - [x] Project Backup includes linked files
 - [x] Project Upgrade works also for linked data
 - [x] Save via linked objects should be optional 

## Summary of Changes

### **1. Writer: Extended linked-object extraction**

- Added support for `aslink="true"`, `"1"`, and `"refonly"`.
- Extracted objects are written into directory structures based on their hierarchical object path.
- `refonly` mode creates an `<objectref>` but does **not** write a linked file.
- Nested linked objects inside linked files are now properly extracted (**recursive extraction**).
- `aslink` attributes are stripped in the external object copies.
- `<object>` nodes in the master document are replaced with `<objectref>` nodes containing:
  - `class`
  - `name`
  - `uuid`
  - `href` (relative to project root)

### **2. Reader: Fully recursive loading of linked objects**

- Implemented `loadProjectXmlWithLinkedObjects()` with full recursion.
- **All `href` paths are resolved relative to the master file directory**, not the directory of each external file.
- Handles arbitrarily deep cascades of linked objects:
  - `master → pkg → leaf → …`
- Cycle protection added to avoid infinite recursion.
- When a linked file cannot be read:
  - A warning is generated
  - The `<objectref>` remains intact
  - Loading continues gracefully

### **3. Project Refactoring**

- Implemented reusable project settings which can be written / read from JSON.
- The settings are stored in the session JSON document, replacing the previous
   boolean only value to persist the irregularity alert settings.
- A new settings is introduced, that steers whether linked object files (new behavior) are used
  or the old behavior of one file per module. This allows to optionally ensure GTlab 2.0 backwards 
  compatibility.

## Edge Case Handling

- If a linked file cannot be resolved (e.g., the file does not exist or cannot be parsed), a *dummy object* is generated and marked as read-only.  
  This allows the project to remain structurally valid without silently dropping the missing object.

- Important: The dummy object does **not** contain the full original data. Re-exporting the project in this state must therefore avoid overwriting the (missing or corrupted) external file with incomplete data.  
  To ensure this, the loader marks such objects with `aslink="refonly"`.  
  This guarantees that when saving:
  - only the `<objectref>` is written back into the master document  
  - **no external `*.gtobj.xml` file is created or overwritten**, preventing accidental data loss.

- If an error occurs during writing of the master or any linked files, the output could otherwise end up in an inconsistent or partially written state.  
  To prevent this, all writes are handled by a `BatchSaver`:
  - Each file is written to a temporary location first  
  - Only if *all* files succeed, they are atomically replaced  
  - Backups are created for every original file before overwriting  
  This ensures that either the entire save succeeds cleanly or no permanent changes are made.


## How Has This Been Tested?

Added an extensive set of tests for both writing *and* loading:

### **Writer tests**
- No `aslink` → only master
- Single linked object
- Multiple linked objects
- `aslink="refonly"`
- Hierarchical extraction
- **Recursive extraction**

### **Reader tests (new)**
- Master-only (no `<objectref>`)
- Single linked object
- Missing linked file → warning and preserved `<objectref>`
- Hierarchical expansion
- **Recursive expansion with nested external files**

Together, these tests verify the full round-trip behavior:

> **Master XML → extraction → objectref substitution → recursive load → full reconstruction**



## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. -->
- [x] A test for the new functionality was added (if applicable).
- [x] All tests run without failure.
- [x] The changelog has been extended, if this MR contains important changes from the users's point of view.
- [x] The new code complies with the GTlab's style guide.
- [x] New interface methods / functions are exported via `EXPORT`. Non-interface functions are NOT exported.
- [x] The number of code quality warnings is not increasing.

rainman110 committed:
Added atomic batch saver

This batch saver is used to atomicly write
multiple files and create backups.

If any of the file write fails, no file
will be written.

Addresses #1419

rainman110 committed:
Added unit tests

rainman110 committed:
Added tests and small refactoring

rainman110 committed:
Fixed cppcheck

rainman110 committed:
Refactoring and first working implementation of linked file use

rainman110 committed:
Enforce a separate dir per package

rainman110 committed:
Improved test readability

rainman110 committed:
Implemented reading of linked objects

rainman110 committed:
Fixed not all warnings printed

rainman110 committed:
Handle case, when objectrefs cannot be restored

 - In this case, a dummy object is created
 - Avoid, recreation of default (dummy) linked data
   when the object is serialized

rainman110 committed:
Fixed writing nested linked objects

rainman110 committed:
Moved project xml saving to data processor

rainman110 committed:
Fixed recursive loading paths

Also added reader tests

rainman110 committed:
Make module upgrade include linked files

rainman110 committed:
Fix project backup not writing linked files

rainman110 committed:
Extended CHANGELOG.md

rainman110 committed:
Some refactoring of batchsaver

rainman110 committed:
Store serialization info to project settings

This required to create a project settings in the first place.
The new projectSettings class is a key value pair and can be
extended to include more project related settings.

rainman110 committed:
Make link file serialization optional (per project)

jensschmeink commented on src/dataprocessor/gt_batchsaver.cpp:
```diff
@@ -0,0 +1,261 @@
+        lastError = msg;
+    }
+
+    bool prepareSaveFiles(const QVector<GtBatchSaver::Op>& ops,
```
Just a question of interest: why do you mix QVector and std::vector in one function-signature?

jensschmeink commented on src/dataprocessor/gt_batchsaver.h:
```diff
@@ -0,0 +1,146 @@
+     *
+     * @param targetPath Target file path for the XML.
+     * @param doc        QDomDocument to be serialized.
+     * @param indent     Indentation width used by QDomDocument::save (default: 4).
```
Indent is no argument but attrOrdered

jensschmeink commented on src/dataprocessor/gt_xmlutilities.cpp:
```diff
@@ -242,3 +285,417 @@ gt::xml::readDomDocumentFromFile(QFile& file,
+}
+
+void
+expandObjectRefsInDocument(QDomDocument& doc,
```
Maybe add a hint, that the implementation canbe found some lines below

jensschmeink reviewed (changes requested):
Hey @rainman110 I added some comments and questoins. I think all of them can be solved quickly. I really enjoy the good unittests!

This is my code review. I haven’t tested it locally yet, but I will do so before giving approval.

rainman110 committed:
Applied proposals from the code review

rainman110 committed:
Fixed dummy object not created, when xml cannot be resolved

rainman110 committed:
Add unresolved linked-object restore edge case tests

rainman110 committed:
Consolidate linked-object restore tests and cross-platform fixes

rainman110 committed:
Fix BatchSaver binary writes to use non-text file mode

rainman110 committed:
Propagate linked-save errors and implement project settings cend

rainman110 committed:
Document SaveAsOwnFile linked XML serialization

rainman110 committed:
Rework data modeling docs for mementos and serialization

jensschmeink reviewed (approved):
Now it looks fine to me

rainman110 commented:
@real-ct-ac We have jens approval, it seems to work. I still like your opinion here with this BIG change in behaviour.

jensschmeink commented:
> @real-ct-ac We have jens approval, it seems to work. I still like your opinion here with the BIG change.

@rainman110 What is this idea of a BIG change about?

rainman110 commented:
> @rainman110 What is this idea of a BIG change about?

I meant, that this is a big change in behaviour of the data processor and i'd like to have a second opinion.

rainman110 commented:
@copilot resolve the merge conflicts in this pull request

rainman110 committed:
Merge origin/master

rainman110 committed:
Fix cppcheck warning in project settings tab

rainman110 committed:
Added tests for new regressions

1. Older session files store projects[path] as a plain boolean for IgnoreIrregularities.
GtProjectSettings::fromJson() now returns immediately for any non-object payload,
so loading an existing session silently resets that setting to the default instead
of preserving the user's preference.
This is a regression for anyone upgrading from a pre-change session file.

2. The BatchSaver deletes any existing *_backup before the transaction
is known to succeed. If a later file in the batch fails to rename or commit,
the rollback can restore the current target file, but the previous backup is
already gone, so the last known-good version is lost even though the save aborted.
Delaying backup removal until after a successful commit would avoid that data loss.

rainman110 committed:
Keep previous batch backups

Move existing _backup files aside until the batch succeeds, then remove the temporary copy only after all QSaveFile commits complete. If a later step fails, the rollback path restores the original backup so the last known-good file is preserved.

rainman110 committed:
Restore legacy session settings

Accept the old top-level boolean session payload as the irregularity flag before falling back to object parsing. That keeps pre-change session files readable instead of silently resetting the project preference to its default.

rainman110 committed:
Add regression tests for coverage gaps

Cover legacy project settings JSON handling, batch saver rollback, project filename helpers, and XML utility branches reported by Codecov.

rainman110 committed:
Improve test coverage for project I/O

rainman110 committed:
Add diff coverage target to CodeCoverage

rainman110 committed:
Make diff-cover tool optional

rainman110 committed:
Make gcovr coverage targets optional
